### PR TITLE
Introduced ContainersCollection

### PR DESCRIPTION
**Pull request by nulls**
### What's done:
- hide all logic is related to container's info to a dedicated class

### CODE REVIEW

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/SaveOrchestratorCommonConfiguration.kt [hunk 1]
```diff
@@ -15,4 +16,8 @@ import org.springframework.scheduling.annotation.EnableScheduling
 @EnableConfigurationProperties(ConfigProperties::class)
 @EnableScheduling
 @PropertySource("classpath:META-INF/save-orchestrator-common/application.properties")
-class SaveOrchestratorCommonConfiguration
+class SaveOrchestratorCommonConfiguration {
+    fun containersCollection(configProperties: ConfigProperties): ContainersCollection {
+        return ContainersCollection(configProperties.agentsHeartBeatTimeoutMillis)
```
Comment (nulls) [inline save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/SaveOrchestratorCommonConfiguration.kt:21]: it's not needed

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt [hunk 2]
```diff
@@ -104,32 +106,33 @@ class AgentService(
...
             }
-            .filter { (_, finishedContainerIds) -> finishedContainerIds.isNotEmpty() }
-            .flatMap { (executionId, finishedContainerIds) ->
+            .filter { finishedContainerIds -> finishedContainerIds.isNotEmpty() }
+            .flatMap { finishedContainerIds ->
                 log.info { "For execution id=$executionId all agents have completed their lifecycle" }
                 markExecutionBasedOnAgentStates(executionId, finishedContainerIds)
-                    .thenReturn(
+                    .then(Mono.fromCallable {
+                        orchestratorAgentStatusService.deleteAllByExecutionId(executionId)
```
Comment (kgevorkyan) [inline save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt:129]: May be rename this method, thus to point that we actually clear our collections and not removing containers from `db` or `docker`? As I remember, now `cleanup `is responsible for this

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt [hunk 3]
```diff
@@ -104,32 +106,33 @@ class AgentService(
...
                         containerRunner.cleanupAllByExecution(executionId)
-                    )
+                    })
             }
             .doOnSuccess {
                 if (it == null) {
-                    log.debug("Agents other than $containerId are still running, so won't try to stop them")
+                    log.debug("Agents for execution $executionId are still running, so won't try to stop them")
```
Comment (kgevorkyan) [inline save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt:135]: We cant determine now which concrete containers `are still running`?

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/ContainerService.kt [hunk 4]
```diff
@@ -85,31 +81,22 @@ class ContainerService(
                     interval = configProperties.agentsStartCheckIntervalMillis.milliseconds,
                     numberOfChecks = configProperties.agentsStartTimeoutMillis / configProperties.agentsStartCheckIntervalMillis
                 ) {
-                    areAgentsHaveStarted.computeIfAbsent(executionId) { AtomicBoolean(false) }.get()
+                    !orchestratorAgentStatusService.containsAnyByExecutionId(executionId)
```
Comment (kgevorkyan) [inline save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/ContainerService.kt:84]: Also, can we rename this method is some manner it, to point out that we address here to our collection and not to `db`?

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/controller/HeartbeatController.kt [hunk 5]
```diff
@@ -56,7 +56,6 @@ class HeartbeatController(private val agentService: AgentService,
         val containerId = heartbeat.agentInfo.containerId
         logger.info("Got heartbeat state: ${heartbeat.state.name} from $containerId under execution id=$executionId")
         return {
-            containerService.markAgentForExecutionAsStarted(executionId)
```
Comment (kgevorkyan) [inline save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/controller/HeartbeatController.kt:59]: now we don't need any initialization? of collection maybe

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt (hunk 3 above)
Comment (nulls) [reply save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt:135]: no, we can't here. this method only waits that all agents are stopped by changing state in database

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt (hunk 2 above)
Comment (nulls) [reply save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt:129]: actually, I did it on purposes -- `HeartBeatInspector` with info about containers mimics\repeats `AgentStatusRepository` from backend, but now it's in-memory.

The class and field has name with prefix `orchestrator` to highlight that it's on orchestrator level

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/ContainerService.kt (hunk 4 above)
Comment (nulls) [reply save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/ContainerService.kt:84]: I wanted to highlight it by prefix `orchestrator`. Can rename it to `agentStatusInMemoryRepository`

Comment (nulls) [reply save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/ContainerService.kt]: renamed the class and field name

File: save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt (hunk 2 above)
Comment (nulls) [reply save-orchestrator-common/src/main/kotlin/com/saveourtool/save/orchestrator/service/AgentService.kt]: renamed the class and field name

**Review by kgevorkyan — APPROVED** [review]
lgtm

**Pull request merged**